src/app/utils/pickupTimeEstimator.js

```diff
@@ -107,28 +104,74 @@ estimator._addOnsiteTravelDuration = async (serviceTime, locationId) => {
   return estimator._addDuration(serviceTime, onsiteTravelDuration)
 }
 
-
 /**
- *  Returns true if the named location has a special delivery schedule (e.g. is
- *  a scholar room)
- *
- *  TODO: Implement
- */
-estimator._hasSpecialDeliverySchedule = (locationId) => {
-  // Return true if locationId matches room known to have a specific delivery schedule
-  return false
-}
-
-/**
- *  Given a location id and a timestamp string, returns a new timestamp string
- *  representing the future delivery time for the location given the known
- *  delivery schedule.
- *
- *  TODO: Implement
+ * Given a location id and a timestamp string, returns an object with:
+ * hasSpecialDeliverySchedule: a Boolean which is true if the location matches a
+ * location with a special delivery schedule
+ * adjustedSpecialScheduleTime: a timestamp string, representing the future delivery time for
+ * the given known delivery schedule
  */
 estimator._adjustToSpecialSchedule = (locationId, time) => {
-  // Update time to next delivery schedule after given time
-  return time
+  let hasSpecialDeliverySchedule = false
+  let adjustedSpecialScheduleTime = new Date(time)
+  let secondFloorScholarRooms = ['mal17', 'mala', 'malc', 'maln', 'malw']
+  let mapRooms = ['mapp8', 'mapp9', 'map08']
+  let firstHour
+  let getNextHour
+  let lastHour
+
+  if (secondFloorScholarRooms.includes(locationId)) {
+    hasSpecialDeliverySchedule = true
+    firstHour = 10
+    lastHour = 16
+    getNextHour = hour => 2*(parseInt(hour/2 + 1))
+  }
+
+  if (mapRooms.includes(locationId)) {
+    hasSpecialDeliverySchedule = true
+    firstHour = 11
+    lastHour = 15
+    getNextHour = hour => 2*(parseInt(hour/2 + 0.5)) + 1
+  }
+
+  if (hasSpecialDeliverySchedule) {
+    adjustedSpecialScheduleTime.setMilliseconds(
+      adjustedSpecialScheduleTime.getMilliseconds() - 1
+    )
+
+    let nextHour = getNextHour(adjustedSpecialScheduleTime.getHours())
+    adjustedSpecialScheduleTime.setHours(nextHour)
+    adjustedSpecialScheduleTime.setMinutes(0)
+    adjustedSpecialScheduleTime.setSeconds(0)
+    adjustedSpecialScheduleTime.setMilliseconds(0)
+
+    if (adjustedSpecialScheduleTime.getHours() < firstHour) {
+      adjustedSpecialScheduleTime.setHours(firstHour)
+    }
+
+    if (adjustedSpecialScheduleTime.getHours() > lastHour) {
+      adjustedSpecialScheduleTime.setDate(
+        adjustedSpecialScheduleTime.getDate() + 1
+      )
+      adjustedSpecialScheduleTime.setHours(firstHour)
+    }
+
+    let day = adjustedSpecialScheduleTime.getDay()
+    if (day === 0 || day === 6) {
+      let daysToIncrement = (8 - day) % 7
+      adjustedSpecialScheduleTime.setDate(
+        adjustedSpecialScheduleTime.getDate() + daysToIncrement
+      )
+      adjustedSpecialScheduleTime.setHours(firstHour)
+    }
```

Comment:
It seems like this assumes Sat and Sun closure and maybe doesn't handle end-of-month rollover? (i.e. if the initial adjusted date is Sunday, the 31st, will this attempt to `setDate(32)`?

I think we can use the [getNextServiceTime](https://github.com/NYPL/discovery-front-end/blob/15c3dd8ca7d2a88e4c40d338b65bcc30ef763100/src/app/utils/pickupTimeEstimator.js#L319) call to roll the adjusted date forward like so:

```
if (adjustedSpecialScheduleHours > lastHour) {
  // Bump the current adjusted time forward by a couple hours to ensure we exceed closing:
  const adjustedTimestamp = estimator.addDuration(
    adjustedSpecialScheduleTime.toISOString(),
    'PT2H'
  )
  // When will the building be open next?
  const nextBuildingHours = estimator._getNextServiceHours(adjustedTimestamp)
  // Now we can just recurse:
  return estimator._adjustToSpecialSchedule(nextBuildingHours.startTime)
...
```